### PR TITLE
Add pause library, make Mailbox pausable and with reentrancy guard

### PR DESCRIPTION
**Pull request by tkporter**
* Adds `pause` library. Implemented it in a similar fashion to https://github.com/FuelLabs/sway-libs/tree/master/libs/ownership/src
* Implements Pausable for the Mailbox contract, which lets the owner pause the contract
* Makes `dispatch` revert if the contract is paused
* Makes process revert if the contract is paused or there is reentrancy

### CODE REVIEW

File: contracts/pause-test/tests/harness.rs [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+
+#[tokio::test]
+async fn test_pause() {
+    let (contract, _id) = get_contract_instance().await;
+
+    // Pause
+    let call = contract.methods().pause().call().await.unwrap();
+
+    // Expect an event
+    let events = call.get_logs_with_type::<PausedEvent>().unwrap();
```
Comment (yorhodes) [inline contracts/pause-test/tests/harness.rs:59]: ahh this is a nice feature :)

File: contracts/pause/src/lib.sw [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+    match get_pause_state() {
+        PausedState::Unpaused => false,
+        PausedState::Paused => true,
+    }
+}
+
+/// Gets the PausedState from storage.
+#[storage(read)]
+fn get_pause_state() -> PausedState {
+    get(PAUSED_STORAGE_KEY).unwrap_or(PausedState::Unpaused)
```
Comment (yorhodes) [inline contracts/pause/src/lib.sw:33]: this default is a little scary to me

File: contracts/pause/src/lib.sw [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+    require(!is_paused(), "contract is paused");
+}
+
+/// Returns whether the contract is paused.
+#[storage(read)]
+pub fn is_paused() -> bool {
+    match get_pause_state() {
+        PausedState::Unpaused => false,
+        PausedState::Paused => true,
+    }
```
Comment (yorhodes) [inline contracts/pause/src/lib.sw:27]: why not just use a bool?

File: contracts/hyperlane-mailbox/tests/harness.rs [hunk 4]
```diff
@@ -282,6 +313,159 @@ async fn test_latest_checkpoint() {
...
+    let process_call = mailbox
+        .methods()
+        .process(metadata.clone(), agent_message.clone().into())
+        .set_contract_ids(&contract_inputs)
+        .tx_params(TxParameters::default().set_gas_limit(1_200_000))
+        .call()
+        .await
+        .unwrap();
+
+    let message_id = &process_call.get_logs_with_type::<Bits256>().unwrap()[0];
```
Comment (yorhodes) [inline contracts/hyperlane-mailbox/tests/harness.rs:338]: did you mean for this to be in https://github.com/hyperlane-xyz/fuel-contracts/pull/75 ?

File: contracts/hyperlane-mailbox/tests/harness.rs [hunk 5]
```diff
@@ -282,6 +313,159 @@ async fn test_latest_checkpoint() {
...
+        .process(metadata.clone(), agent_message.clone().into())
+        .set_contract_ids(&contract_inputs)
+        .tx_params(TxParameters::default().set_gas_limit(1_200_000))
+        .call()
+        .await
+        .unwrap();
+
+    let msg_recipient = TestMessageRecipient::new(recipient_id, mailbox.account());
+    let handled = msg_recipient.methods().handled().simulate().await.unwrap();
+    assert!(handled.value);
```
Comment (yorhodes) [inline contracts/hyperlane-mailbox/tests/harness.rs:365]: nice

File: contracts/hyperlane-mailbox/tests/harness.rs [hunk 6]
```diff
@@ -282,6 +313,159 @@ async fn test_latest_checkpoint() {
...
+    let process_delivered_error = mailbox
+        .methods()
+        .process(metadata.clone(), agent_message.clone().into())
+        .set_contract_ids(&contract_inputs)
+        .tx_params(TxParameters::default().set_gas_limit(1_200_000))
+        .call()
+        .await
+        .unwrap_err();
+
+    assert_eq!(get_revert_string(process_delivered_error), "delivered");
```
Comment (yorhodes) [inline contracts/hyperlane-mailbox/tests/harness.rs:407]: glad you revisited these :)

**Review by yorhodes — COMMENTED** [review]
should we add a test for reentrant message recipient?

**Comment by yorhodes** [comment]
just stumbled upon https://github.com/FuelLabs/sway-libs/issues/102

File: contracts/pause/src/lib.sw (hunk 2 above)
Comment (tkporter) [reply contracts/pause/src/lib.sw:33]: the alternative is an explicit initializer, wdyt?

imo this is well tested so it's pretty much the same as in Solidity where the `paused` storage variable would be implicitly initialized to its zero value, which means that it's not paused

File: contracts/pause/src/lib.sw (hunk 3 above)
Comment (tkporter) [reply contracts/pause/src/lib.sw:27]: great point lmao

**Comment by tkporter** [comment]
> just stumbled upon [FuelLabs/sway-libs#102](https://github.com/FuelLabs/sway-libs/issues/102)

interesting
Think we're fine in this case though as none of the other contracts depend on stuff like whether a specific message has been delivered or not

File: contracts/hyperlane-mailbox/tests/harness.rs (hunk 4 above)
Comment (tkporter) [reply contracts/hyperlane-mailbox/tests/harness.rs]: yeah this is gone now

**Review by yorhodes — APPROVED** [review]

**Pull request merged**